producer/writer/options.go

```diff
@@ -0,0 +1,361 @@
+)
+
+const (
+	defaultDialTimeout               = 10 * time.Second
+	defaultMessageRetryDelay         = 5 * time.Second
+	defaultAckErrRetryDelay          = 10 * time.Second
+	defaultPlacementWatchRetryDelay  = 5 * time.Second
+	defaultPlacementWatchInitTimeout = 5 * time.Second
+	defaultTopicWatchInitTimeout     = 5 * time.Second
+	defaultCloseCheckInterval        = 2 * time.Second
+	defaultConnectionResetDelay      = 2 * time.Second
+	// Using 16K which provides much better performance comparing
+	// to lower values like 1k ~ 8k.
+	defaultBufferSize = 16384
+)
+
+// Options configs the writer.
+type Options interface {
+	// TopicName returns the topic name.
+	TopicName() string
+
+	// SetTopicName sets the topic name.
+	SetTopicName(value string) Options
+
+	// TopicService returns the topic service.
+	TopicService() topic.Service
+
+	// SetTopicService sets the topic service.
+	SetTopicService(value topic.Service) Options
+
+	// TopicWatchInitTimeout returns the timeout for topic watch initialization.
+	TopicWatchInitTimeout() time.Duration
+
+	// SetTopicWatchInitTimeout sets the timeout for topic watch initialization.
+	SetTopicWatchInitTimeout(value time.Duration) Options
+
+	// ServiceDiscovery returns the client to service discovery service.
+	ServiceDiscovery() services.Services
+
+	// SetServiceDiscovery sets the client to service discovery services.
+	SetServiceDiscovery(value services.Services) Options
+
+	// PlacementWatchRetryDelay returns the delay before retrying on placement watch errors.
+	PlacementWatchRetryDelay() time.Duration
+
+	// SetPlacementWatchRetryDelay sets the delay before retrying on placement watch errors.
+	SetPlacementWatchRetryDelay(value time.Duration) Options
+
+	// PlacementWatchInitTimeout returns the timeout for placement watch initialization.
+	PlacementWatchInitTimeout() time.Duration
+
+	// SetPlacementWatchInitTimeout sets the timeout for placement watch initialization.
+	SetPlacementWatchInitTimeout(value time.Duration) Options
+
+	// MessagePoolOptions returns the options of pool for messages.
+	MessagePoolOptions() pool.ObjectPoolOptions
+
+	// SetMessagePoolOptions sets the options of pool for messages.
+	SetMessagePoolOptions(value pool.ObjectPoolOptions) Options
+
+	// MessageRetryBackoff returns the backoff before retrying messages.
+	MessageRetryBackoff() time.Duration
+
+	// SetMessageRetryBackoff sets the backoff before retrying messages.
+	SetMessageRetryBackoff(value time.Duration) Options
+
+	// CloseCheckInterval returns the close check interval.
+	CloseCheckInterval() time.Duration
+
+	// SetCloseCheckInterval sets the close check interval.
+	SetCloseCheckInterval(value time.Duration) Options
+
+	// AckErrorRetryDelay returns the delay before retring on ack errors.
+	AckErrorRetryDelay() time.Duration
+
+	// SetAckErrorRetryDelay sets the delay before retring on ack errors.
+	SetAckErrorRetryDelay(value time.Duration) Options
+
+	// DialTimeout returns the dial timeout.
```

Comment:
nit: Create a separate `ConnectionOptions` struct and move these connection related options to that type.